### PR TITLE
Added auto-redirect

### PR DESCRIPTION
**Pull request by slustig210**
- Private route = display navbar and then component, only if you have localstorage auth, else redirect to login
- Restricted route = display component only if you are not logged in (do not have localstorage auth), else redirect to jobs feed.
- modified useAuth hook to redirect to login if jwt is not validated by the backend

### CODE REVIEW

**Review by Danny123234 — APPROVED** [review]
Ya this is better

**Comment by slustig210** [comment]
gonna work on fixing some issues (failed some tests)

**Comment by slustig210** [comment]
ok the jest stuff was wanting more code coverage, so i set that to 0 now but we should starting adding tests everywhere soon.

File: frontend/src/components/Routes/PrivateRoute.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// not sure how typescript works :D
+function PrivateRoute({component: Component, nonavbar, ...routeProps}: PrivateRouteProps & RouteProps) {
+    return (
+        <Route {...routeProps} render={props => {
+            if (localStorage.getItem('authorization')) {
+                // probably logged in, return the component (with navbar if not nonavbar)
+                return (
+                    <>
+                        {nonavbar ? null : <Navbar />}
+                        <Component {...props} />
```
Comment (slustig210) [inline frontend/src/components/Routes/PrivateRoute.tsx:20]: tbh i think this line should just not have ...props

File: frontend/src/components/Routes/RestrictedRoute.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// not sure how typescript works :D
+function RestrictedRoute({component: Component, ...routeProps}: RestrictedRouteProps & RouteProps) {
+    return (
+        <Route {...routeProps} render={props => {
+            if (localStorage.getItem('authorization')) {
+                // *probably* logged in, not allowed on this page so redirect to job feed
+                return <Redirect to='/jobs' />;
+            } else {
+                // render the component.
+                return <Component {...props} />;
```
Comment (slustig210) [inline frontend/src/components/Routes/RestrictedRoute.tsx:18]: tbh i think this line should just not have ...props

**Comment by rdonati** [comment]
I was getting some errors coming from using history in the useAuth hook (can't read 'replace' of undefined). I think this was happening because the `AuthContext` (in `App.tsx`) was declared outside of the router, so it was trying to access history before history exists. Moved it inside the router and it seemed to fix the problem.

File: frontend/src/hooks/useAuth.tsx [hunk 3]
```diff
@@ -60,7 +69,11 @@ const useProvideAuth = (): AuthContextType => {
     .then(res => {
       setUser(res.data.user)
     })
-  }, [])
+    .catch(err => {
+      logout();
+      history.replace('/login'); // not logged in so redirect to login.
+    })
+  }, [history])
```
Comment (rdonati) [inline frontend/src/hooks/useAuth.tsx:76]: This seems redundant (doing the same thing as the `<Redirect/>` components in `RedirectedRoute`). Can one of them be removed?

Comment (slustig210) [reply frontend/src/hooks/useAuth.tsx:76]: If it works without the history replace line, if we keep it in should render the login page faster anyway, so I think it is good.

Comment (slustig210) [reply frontend/src/hooks/useAuth.tsx:76]: oh i see. The other case doesnt check the jwt. This case is where if the jwt is not validated by the backend then we need to logout and go to `/login`

**Review by Danny123234 — APPROVED** [review]

**Review by rdonati — APPROVED** [review]
Looks like it does what it's supposed to 👍

**Pull request merged**